pyrit/orchestrator/fuzzer_orchestrator.py

```diff
@@ -0,0 +1,427 @@
+            key = self._best_UCT_score()
+        )
+        self._mctc_select_path.append(current)
+
+        # while node is not a leaf
+        while len(current._children) > 0:  
+            if np.random.rand() < self._non_leaf_nodeprobability:
+                break
+            #compute the bestUCT score
+            current = max(  
+                current._children,
+                key= self._best_UCT_score()
+            )
+            # append node to path
+            self._mctc_select_path.append(current) 
+
+        for prompt_node in self._mctc_select_path:
+            # keep track of number of visited nodes
+            prompt_node._visited_num += 1    
+
+        self._last_choice_index = current.index
+        # returns the best child
+        return current 
+
+    def _best_UCT_score(self):
+        
+        """Function to compute the score for each seed. The highest-scoring seed will be selected as the next seed.  
+        
+         This is an extension of the Monte-carlo tree search algorithm which applies Upper Confidence Bound (UCB) to the node. 
+         
+         UCB function determines the confidence interval for each node and returns the highest value which will be selected as next seed."""
+        
+        return lambda pn: self._rewards[pn.index] / (pn._visited_num + 1) +  \
+            self._frequency_weight * np.sqrt(2 * np.log(self._step) /  (pn._visited_num + 0.01)) # self._frequency_weight - constant that balances between the seed with high reward and the seed that is selected fewer times. 
+       
+
+    def _update(self, prompt_nodes: 'PromptNode'): 
+        """
+        Updates the reward of all the nodes in the last chosen path. 
+        """
+        success_number = self._num_jailbreak #check 
+
+        last_chosen_node = self._prompt_nodes[self._last_choice_index]
+        for prompt_node in reversed(self._mctc_select_path):
+            #the output from the template converter will always be a single template so the length of prompt_nodes will be 1.
+            reward = success_number / (len(self._prompts)
+                                 * 1) # len(prompt_nodes))  
+            self._rewards[prompt_node.index] += reward * \
+                max(self._minimum_reward, (1 - self._reward_penalty * last_chosen_node._level))
+            
+    @pyrit_placeholder_retry
+    async def _apply_template_converter(self,current_seed):
+        """
+        Asynchronously applies template converter.
+
+        Retries the function if it raises MissingPromptHolderException,
+        Logs retry attempts at the INFO level and stops after a maximum number of attempts.
+
+        Args:
+            current_seed: The template that is selected. 
+
+        Returns:
+            converted template with placeholder for prompt. 
+
+        Raises:
+            MissingPromptHolderException: If the prompt placeholder is empty after exhausting the maximum number of retries.

```

Comment:
```suggestion
            MissingPromptHolderException: If the prompt placeholder is still missing.
```
The method is not aware of retries. Those happen via the decorator.